In [ ]:
!pip install python-dotenv
import pandas as pd
import os
from dotenv import load_dotenv
from sqlalchemy import create_engine

In [ ]:
# Load environment variables from .env file
load_dotenv()

True

In [ ]:
# Accessing the variables
user = os.getenv("DB_USER")
password = os.getenv("DB_PASSWORD")
host = os.getenv("DB_HOST")
port = os.getenv("DB_PORT")
database = os.getenv("DB_NAME")

In [ ]:
# Test loading the environment variables 
print(f"DB_USER: {os.getenv('DB_USER')}")
print(f"DB_HOST: {os.getenv('DB_HOST')}")
print(f"DB_PORT: {os.getenv('DB_PORT')}")
print(f"DB_NAME: {os.getenv('DB_NAME')}")

DB_USER: postgres
DB_HOST: isba-dev-02.cxcoakgsc7ha.us-east-1.rds.amazonaws.com
DB_PORT: 5432
DB_NAME: postgres


In [ ]:
import pandas as pd
import os

# Adjusted path: going one directory up to access data/
file_path = os.path.join(os.getcwd(), '..', 'data', 'local_chatbot_dataset.csv')

# Load the CSV file using the correct path
df = pd.read_csv(file_path)

# Preview the first few rows to confirm it loaded
df.head()


ID               Model Name     Category      Domain        Query Type  \
0   1             Lin-BurnsBot       Hybrid   Education     Complex Query   
1   2  BrownMillerandNguyenBot  Open-Source   Education  Multi-turn Query   
2   3      Trevino-PhillipsBot       Hybrid     Finance      Simple Query   
3   4  SnyderWarrenandBrownBot   Rule-Based   Education      Simple Query   
4   5         Crawford-HoodBot  Open-Source  Healthcare      Simple Query   

   Latency (ms)  Response Accuracy (%)  Performance Time (s)  \
0           266                  80.83                  4.52   
1           698                  91.80                  3.27   
2           285                  91.26                  3.14   
3           108                  83.11                  3.03   
4           759                  81.56                  3.07   

   Memory Usage (MB)  Concurrent Queries      Optimization Technique  \
0                817                  35               Model Pruning   
1               1493                  26  Neural Architecture Search   
2               1414                  21      Reinforcement Learning   
3                937                  26               Model Pruning   
4               1877                  40  Neural Architecture Search   

                        Notes  
0  Improved adaptive response  
1         Faster architecture  
2         Faster architecture  
3         Faster architecture  
4     No optimization applied

In [ ]:
import os
from sqlalchemy import create_engine
import pandas as pd

# Load environment variables (replace with your actual values if needed)
user = os.getenv("DB_USER")
password = os.getenv("DB_PASSWORD")
host = os.getenv("DB_HOST")
port = os.getenv("DB_PORT")
database = os.getenv("DB_NAME")

# Create the connection string and engine
connection_string = f'postgresql+psycopg2://{user}:{password}@{host}:{port}/{database}'
print("Connection String:", connection_string)

# Create the engine and a raw connection
try:
    engine = create_engine(connection_string)
    connection = engine.raw_connection()
    print("Successfully connected to the database using raw connection")
except Exception as e:
    print(f"Connection failed: {e}")


Connection String: postgresql+psycopg2://postgres:isba_4715@isba-dev-02.cxcoakgsc7ha.us-east-1.rds.amazonaws.com:5432/postgres


Successfully connected to the database using raw connection


In [ ]:
# Listing all tables in the database to confirm your table exists
table_check_query = '''
SELECT table_name 
FROM information_schema.tables 
WHERE table_schema = 'public';
'''

tables = pd.read_sql(table_check_query, engine)
print("Available Tables in Database:")
print(tables)


Available Tables in Database:
                          table_name
0                   Chatbot iOS Apps
1       chatbot_descriptive_analysis
2                chatbot_performance
3  13 Best Free and Paid AI Chatbots


Business Question: Descriptive Analysis Query 
Which chatbot models have the highest average response accuracy, and how do their response times (latency), performance time, and memory usage compare?


In [ ]:
# Defining the descriptive analytics query
sql_query = '''
SELECT 
    "Model Name", 
    AVG("Response Accuracy (%)") AS avg_accuracy,
    AVG("Latency (ms)") AS avg_latency,
    AVG("Performance Time (s)") AS avg_performance_time,
    AVG("Memory Usage (MB)") AS avg_memory_usage
FROM 
    chatbot_performance
GROUP BY 
    "Model Name"
ORDER BY 
    avg_accuracy DESC;
'''

# Debugging: Make sure this is a string and connection is valid
print("SQL Query Type:", type(sql_query))
print("SQL Query:", sql_query)
print("Engine Type:", type(engine))
print("Connection Type:", type(connection))

# Safely running the SQL query using pd.read_sql with raw connection
try:
    df_descriptive = pd.read_sql(sql_query, con=connection)
    print("Query executed successfully")
    pd.set_option('display.max_rows', None)
    df_descriptive.head()
except Exception as e:
    print(f"Query failed: {e}")


SQL Query Type: <class 'str'>
SQL Query: 
SELECT 
    "Model Name", 
    AVG("Response Accuracy (%)") AS avg_accuracy,
    AVG("Latency (ms)") AS avg_latency,
    AVG("Performance Time (s)") AS avg_performance_time,
    AVG("Memory Usage (MB)") AS avg_memory_usage
FROM 
    chatbot_performance
GROUP BY 
    "Model Name"
ORDER BY 
    avg_accuracy DESC;

Engine Type: <class 'sqlalchemy.engine.base.Engine'>
Connection Type: <class 'sqlalchemy.pool.base._ConnectionFairy'>
Query executed successfully


/tmp/ipykernel_1605/2520820109.py:25: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_descriptive = pd.read_sql(sql_query, con=connection)


In [ ]:
import os
from sqlalchemy import create_engine
import pandas as pd

# Load environment variables (replace with your actual values if needed)
user = os.getenv("DB_USER")
password = os.getenv("DB_PASSWORD")
host = os.getenv("DB_HOST")
port = os.getenv("DB_PORT")
database = os.getenv("DB_NAME")

# Create the connection string and engine
connection_string = f'postgresql+psycopg2://{user}:{password}@{host}:{port}/{database}'
engine = create_engine(connection_string)
print("Successfully connected to PostgreSQL using SQLAlchemy engine")


Successfully connected to PostgreSQL using SQLAlchemy engine


In [ ]:
# Save the descriptive analysis results as a PostgreSQL table
try:
    df_descriptive.to_sql(
        'chatbot_descriptive_analysis',  # Table name
        con=engine,                     # Use the engine (SQLAlchemy Engine)
        if_exists='replace',            # Replace table if it already exists
        index=False                     # Don't save index column
    )
    print("Descriptive Analysis results saved to PostgreSQL table: chatbot_descriptive_analysis")
except Exception as e:
    print(f"Failed to save the table: {e}")


Descriptive Analysis results saved to PostgreSQL table: chatbot_descriptive_analysis


In [ ]:
# Displaying the entire 'chatbot_descriptive_analysis' table
sql_query_display = '''
SELECT * 
FROM chatbot_descriptive_analysis;
'''

# Running the query to load the table into a DataFrame
df_display = pd.read_sql(sql_query_display, con=engine)

# Displaying the table
pd.set_option('display.max_rows', None)  # Display all rows
df_display.head()


Model Name  avg_accuracy  avg_latency  avg_performance_time  \
0  KnightMcdowellandMarksBot     89.800000   482.166667              3.146667   
1            Carson-BurkeBot     89.641667   595.333333              2.960000   
2          Cochran-HarrisBot     88.908182   548.181818              3.025455   
3                TapiaPLCBot     88.885000   812.500000              2.980000   
4        Trevino-PhillipsBot     87.941250   520.125000              2.688750   

   avg_memory_usage  
0       1020.333333  
1       1106.583333  
2       1461.636364  
3        977.000000  
4       1367.750000

Insight
- The chatbot model "KnightMcdowellandMarksBot" has the highest average response accuracy at 89.8%, making it the most accurate model in this analysis.
- However, this high accuracy comes with a moderately high latency of 482 ms and an average memory usage of 1,020 MB, which may impact performance on lower-end systems.
- Other high-performing models, such as "Carson-BurkeBot" and "Cochran-HarrisBot", also demonstrate strong accuracy (above 88%) but have even higher latency (595 ms and 548 ms, respectively).
- Models with lower accuracy, like "Lewis-CalderonBot" and "Greene-CoxBot", offer much faster response times (638 ms and 608 ms) with lower memory usage, suggesting a trade-off between speed and accuracy.

Recommendation
- For companies looking to integrate chatbots into their web applications:
  - If accuracy is critical (such as in healthcare or finance), prioritize high-accuracy models like "KnightMcdowellandMarksBot", but ensure the server infrastructure can support their higher memory usage.
  - If speed and efficiency are more important (such as in customer support), choose models like "Lewis-CalderonBot" or "Greene-CoxBot", which offer faster response times with reasonable accuracy.
  - Regularly monitor chatbot performance (accuracy, latency, and memory usage) and optimize for your use case.

Prediction
- If a company integrates one of the high-accuracy models, like "KnightMcdowellandMarksBot", into their web application:
  - Users may experience highly accurate responses, but the response time may feel slower due to the higher latency.
  - Memory usage may become a concern if many users are interacting with the chatbot simultaneously, leading to potential performance issues.
  - Upon further analysis, we may find that optimization techniques, like model pruning, could improve speed without significantly sacrificing accuracy.
